asm_utils/src/data_storage.rs

```diff
@@ -0,0 +1,80 @@
+//! Utilites for generating initialization code that stores data objects in memory.
+
+use std::collections::BTreeMap;
+
+use crate::{data_parser::DataValue, utils::next_multiple_of_four};
+
+pub enum SingleDataValue<'a> {
+    Value(u32),
+    LabelReference(&'a String),
+    Offset(&'a String, &'a String),
+}
+
+pub fn store_data_objects<'a>(
+    objects: impl IntoIterator<Item = &'a (String, Vec<DataValue>)> + Copy,
+    mut memory_start: u32,
+    code_gen: &mut dyn FnMut(u32, SingleDataValue) -> Vec<String>,
+) -> (Vec<String>, BTreeMap<String, u32>) {
+    memory_start = ((memory_start + 7) / 8) * 8;
+    let mut current_pos = memory_start;
+    let mut positions = BTreeMap::new();
+    for (name, data) in objects.into_iter() {
+        // TODO check if we need to use multiples of four.
+        let size: u32 = data
+            .iter()
+            .map(|d| next_multiple_of_four(d.size()) as u32)
+            .sum();
+        positions.insert(name.clone(), current_pos);
+        current_pos += size;
+    }
+
+    let code = objects
+        .into_iter()
+        .filter(|(_, data)| !data.is_empty())
```

Comment:
Is this required? Seems like empty data would yield an empty vector in the `flat_map` down below anyway